flow/multiagent_envs/traffic_light_grid.py

```diff
@@ -0,0 +1,260 @@
+
+    Required from env_params: See parent class
+
+    States
+        See parent class
+    Actions
+        See parent class
+    Rewards
+        See parent class
+    Termination
+        See parent class
+
+    """
+
+    def __init__(self, env_params, sim_params, scenario, simulator='traci'):
+        super().__init__(env_params, sim_params, scenario, simulator)
+
+        for p in ADDITIONAL_ENV_PARAMS.keys():
+            if p not in env_params.additional_params:
+                raise KeyError(
+                    'Environment parameter "{}" not supplied'.format(p))
+
+        # number of nearest lights to observe, defaults to 4
+        self.num_local_lights = env_params.additional_params.get(
+            "num_local_lights", 4)
+
+        # number of nearest edges to observe, defaults to 4
+        self.num_local_edges = env_params.additional_params.get(
+            "num_local_edges", 4)
+
+    @property
+    def observation_space(self):
+        """State space that is partially observed.
+
+        Velocities, distance to intersections, edge number (for nearby
+        vehicles) from each direction, local edge information, and traffic
+        light state.
+        """
+        tl_box = Box(
+            low=0.,
+            high=1,
+            shape=(3 * 4 * self.num_observed +
+                   2 * self.num_local_edges +
+                   3 * (1 + self.num_local_lights),
+                   ),
+            dtype=np.float32)
+        return tl_box
+
+    @property
+    def action_space(self):
+        """See class definition."""
+        if self.discrete:
+            return Discrete(2)
+        else:
+            return Box(
+                low=-1,
+                high=1,
+                shape=(1,),
+                dtype=np.float32)
+
+    def get_state(self):
+        """Observations for each traffic light agent.
+
+        :return: dictionary which contains agent-wise observations as follows:
+        - For the self.num_observed number of vehicles closest and incoming
+        towards traffic light agent, gives the vehicle velocity, distance to
+        intersection, edge_number, density traffic light state.
+        - For edges in the network, gives the density and average velocity.
+        - For the self.num_local_lights number of nearest lights (itself
+        included), gives the traffic light information, including the last
+        change time, light direction (i.e. phase), and a currently_yellow flag.
+        """
+        # Normalization factors
+        max_speed = max(
+            self.k.scenario.speed_limit(edge)
+            for edge in self.k.scenario.get_edge_list())
+        grid_array = self.net_params.additional_params["grid_array"]
+        max_dist = max(grid_array["short_length"], grid_array["long_length"],
+                       grid_array["inner_length"])
+
+        # TODO(cathywu) refactor PO_TrafficLightGridEnv with convenience
```

Comment:
Can you convert this into an issue if you're not going to do it? Same goes for the other TODOs/FIXME. In particular, the "adding local observability" might be a "good first issue".